internal/cmd/io_binary.go

```diff
@@ -0,0 +1,236 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"net/url"
+	"os/exec"
+	"sync"
+	"time"
+
+	"github.com/Microsoft/go-winio"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/pkg/errors"
+)
+
+const (
+	binaryPipeFmt        = `\\.\pipe\binary-%s-%s`
+	binaryCmdWaitTimeout = 5 * time.Second
+)
+
+func newBinaryCmd(ctx context.Context, uri *url.URL, id string, ns string) *exec.Cmd {
+	var args []string
+	for k, vs := range uri.Query() {
+		args = append(args, k)
+		if len(vs) > 0 && vs[0] != "" {
+			args = append(args, vs[0])
+		}
+	}
+
+	execPath := uri.Path
+	if execPath == "" {
+		execPath = uri.Host
+	}
+
+	cmd := exec.CommandContext(ctx, execPath, args...)
+	cmd.Env = append(cmd.Env,
+		"CONTAINER_ID="+id,
+		"CONTAINER_NAMESPACE="+ns,
+	)
+
+	return cmd
+}
+
+type binaryIO struct {
+	cmd *exec.Cmd
+
+	binaryCloser sync.Once
+
+	stdin, stdout, stderr string
+
+	sout, serr io.ReadWriteCloser
+	soutCloser sync.Once
+}
+
+// NewBinaryIO starts a binary logger
+func NewBinaryIO(ctx context.Context, id string, uri *url.URL) (_ UpstreamIO, err error) {
+	ns, err := namespaces.NamespaceRequired(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var sout, serr, w io.ReadWriteCloser
+
+	bio := &binaryIO{}
+
+	stdoutPipe := fmt.Sprintf(binaryPipeFmt, id, "stdout")
+	sout, err = openNPipe(stdoutPipe)
+	if err != nil {
+		return nil, err
+	}
+	bio.stdout = stdoutPipe
+	bio.sout = sout
+
+	stderrPipe := fmt.Sprintf(binaryPipeFmt, id, "stderr")
+	serr, err = openNPipe(stderrPipe)
+	if err != nil {
+		return nil, err
+	}
+	bio.stderr = stderrPipe
+	bio.serr = serr
+
+	waitPipe := fmt.Sprintf(binaryPipeFmt, id, "wait")
+	w, err = openNPipe(waitPipe)
+	if err != nil {
+		return nil, err
+	}
+
+	defer w.Close()
+
+	cmd := newBinaryCmd(ctx, uri, id, ns)
+	cmd.Env = append(cmd.Env,
+		"CONTAINER_STDOUT="+stdoutPipe,
+		"CONTAINER_STDERR="+stderrPipe,
+		"CONTAINER_WAIT="+waitPipe,
+	)
+
+	bio.cmd = cmd
+
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+
+	b := make([]byte, 1)
+	if _, err := w.Read(b); err != nil && err != io.EOF {
+		return nil, errors.Errorf("failed to start binary logger")
+	}
+
+	return bio, nil
```

Comment:
Can we just make the binaryIO object here during the return instead of adding everything to it during the function? E.g.
```go
return &binaryIO{
    cmd: cmd,
    stdout: stdoutPipe,
    sout: sout, 
    stderr: stdErrPipe,
    serr: serr, 
}, nil
```

Reply:
agreed. this is how I had it in the beginning and now I don't remember why I changed it 😄